### PR TITLE
Moving spans from userland to extension

### PR DESCRIPTION
**Pull request by bwoebi**
### Description

Doing a lot of things:
 - Nearly everything happening in Tracer->flush() is now in the extension
 - Spans are allocated and closed directly instead of their ids
 - Spans now expose duration and start time via methods
 - The root span is (if there is configured to be one) allocated and closed from within the extension - close at the very last moment: RSHUTDOWN

Also note: DRAFT, UNTESTED AND BROKEN.

### Readiness checklist
- [ ] (only for Members) Changelog has been added to the release document.
- [ ] Tests added for this feature/bug.

### Reviewer checklist
- [ ] Appropriate labels assigned.
- [ ] Milestone is set.
- [ ] Changelog has been added to the release document. For community contributors the reviewer is in charge of this task.

### CODE REVIEW

File: ext/php8/ddtrace.c [hunk 1]
```diff
@@ -748,6 +808,22 @@ static PHP_FUNCTION(additional_trace_meta) {
     zval_copy_ctor(return_value);
 }
 
+/* {{{ proto string DDTrace\add_global_tag(string $key, string $value) */
+static PHP_FUNCTION(add_global_tag) {
+    UNUSED(execute_data);
+
+    zend_string *key, *val;
+    if (zend_parse_parameters(ZEND_NUM_ARGS(), "SS", &key, &val) == FAILURE) {
+        RETURN_THROWS();
```
Comment (labbati) [inline ext/php8/ddtrace.c:817]: We should not throw. We can log a message in debug mode and proceed the execution.
In general, we never throw exceptions, so users' applications can keep working even if with tracing not perfectly working

File: ext/php8/ddtrace.c [hunk 2]
```diff
@@ -1249,6 +1325,50 @@ static PHP_FUNCTION(dd_trace_peek_span_id) {
...
+    double finish_time_seconds = 0;
+    if (zend_parse_parameters(ZEND_NUM_ARGS(), "|f", &finish_time_seconds) != SUCCESS) {
+        ddtrace_log_debug("unexpected parameter. the environment variable name must be provided");
+        RETURN_FALSE;
+    }
+
+    // TODO: also prevent closing internal root spans? Would need marker though
+    if (!DDTRACE_G(open_spans_top) || !DDTRACE_G(open_spans_top)->execute_data) {
+        zend_throw_exception(NULL, "There is no user-span on the top of the stack. Cannot close.", 0);
+        RETURN_THROWS();
```
Comment (labbati) [inline ext/php8/ddtrace.c:1357]: Same as above, let's not throw. The maximum that we can accept (but we should avoid it) is that tracing is not working as expected, but users' requests never have to be interrupted because of us.

File: ext/php8/serializer.c [hunk 3]
```diff
@@ -426,6 +426,36 @@ static void _serialize_meta(zval *el, ddtrace_span_fci *span_fci) {
...
+    ZEND_HASH_FOREACH_STR_KEY_VAL(global_tags, global_key, global_val) {
+        zend_hash_add(Z_ARR_P(meta), global_key, global_val);
+    }
+    ZEND_HASH_FOREACH_END();
+    zend_hash_release(global_tags);
+
+    zend_string *tag_key;
+    zval *tag_value;
+    ZEND_HASH_FOREACH_STR_KEY_VAL(DDTRACE_G(additional_global_tags), tag_key, tag_value) {
+        zend_hash_add(DDTRACE_G(additional_global_tags), tag_key, tag_value);
```
Comment (labbati) [inline ext/php8/serializer.c:455]: Not sure I understand what this line does. I think that here we should add from `DDTRACE_G(additional_global_tags)` into `Z_ARR_P(meta)`?

If this holds true, then we are missing a test for global + additional globals?

File: src/DDTrace/Bootstrap.php [hunk 4]
```diff
@@ -51,7 +53,24 @@ public static function tracerOnce()
...
+                        }
+                        /*
+                         * Having this priority sampling here is actually a bug (should happen after service name
+                         * substitutions), but it was this way before the refactor, so let's fix this in a
+                         * subsequent release, when we will have ported everything else to the extension.
+                         */
+                        $tracer = GlobalTracer::get();
+                        if (method_exists($tracer, "enforcePrioritySamplingOnRootSpan")) {
+                            $tracer->enforcePrioritySamplingOnRootSpan();
+                        }
```
Comment (labbati) [inline src/DDTrace/Bootstrap.php:72]: I am not understanding this part. Reaching out in private to have a quick sync.

**Review by labbati — COMMENTED** [review]
Thank you Bob, added a few comments in the meantime. I am also going to test a specific edge case I have in mind and that I think is not covered in a test, but I will report back on that with an additional comment

File: src/DDTrace/Bootstrap.php [hunk 5]
```diff
@@ -126,6 +145,14 @@ private static function initRootSpan(Tracer $tracer)
             foreach (Private_\util_extract_configured_headers_as_tags($httpHeaders, true) as $tag => $value) {
                 $span->setTag($tag, $value);
             }
+
+            if (PHP_VERSION_ID >= 80000) {
+                foreach ($httpHeaders as $header => $value) {
+                    if (stripos($header, Propagator::DEFAULT_ORIGIN_HEADER) === 0) {
```
[the remark this answers is not in the document]
Comment (bwoebi) [reply src/DDTrace/Bootstrap.php:151]: while that's totally possible, it's in line with the current handling: https://github.com/DataDog/dd-trace-php/blob/master/src/DDTrace/Propagators/CurlHeadersMap.php#L33 If we want to change that, we probably should do that separately...?

File: src/DDTrace/Tracer.php [hunk 6]
```diff
@@ -253,7 +274,7 @@ public function startActiveSpan($operationName, $options = [])
             $span->setTag(Tag::SERVICE_NAME, $parentService);
         }
 
-        return $this->scopeManager->activate($span, $options->shouldFinishSpanOnClose());
+        return $this->scopeManager->activate($span, $options->shouldFinishSpanOnClose() && (PHP_VERSION_ID >= 80000 && $span->getParentId() == 0));
```
[the remark this answers is not in the document]
Comment (bwoebi) [reply src/DDTrace/Tracer.php:277]: Yeah that's not right. And there's also a failing test for that. (but as said, this current PR is fully untested, locally I've done some progress.)

File: ext/php8/serializer.c (hunk 3 above)
Comment (bwoebi) [reply ext/php8/serializer.c:455]: Yes - and yes, for sure that'll get a test.

File: src/DDTrace/Bootstrap.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (labbati) [reply src/DDTrace/Bootstrap.php:151]: good point, please create a ticket that we can address later as a low hanging fruit

File: ext/php8/span.h [hunk 7]
```diff
@@ -14,25 +14,23 @@ static const int ddtrace_num_error_tags = 3;
...
     uint64_t trace_id;
     uint64_t parent_id;
     uint64_t span_id;
     uint64_t start;
-    union {
-        uint64_t duration_start;
-        uint64_t duration;
-    };
+    uint64_t duration_start;
+    uint64_t duration;
```
Comment (SammyK) [inline ext/php8/span.h:24]: What is the motivation for removing the `union`?

**Review by SammyK — COMMENTED** [review]
Looking great so far! I'm still reviewing this but I left a few comments/questions so far.

File: ext/php8/span.h (hunk 7 above)
Comment (bwoebi) [reply ext/php8/span.h:24]: Because the userland interface wants to distinguish between finished and unfinished spans; now we can just return duration == 0, then userland knows unfinished

File: ext/php8/ddtrace.c [hunk 8]
```diff
@@ -1249,6 +1327,50 @@ static PHP_FUNCTION(dd_trace_peek_span_id) {
...
 }
 
+/* {{{ proto string dd_trace_peek_span() */
+static PHP_FUNCTION(dd_trace_peek_span) {
+    UNUSED(execute_data);
+    if (DDTRACE_G(open_spans_top)) {
+        RETURN_OBJ_COPY(&DDTRACE_G(open_spans_top)->span.std);
+    }
+    RETURN_NULL();
+}
```
[the remark this answers is not in the document]
Comment (bwoebi) [reply ext/php8/ddtrace.c:1337]: The span is is currently not exposed on SpanData, but we may add getters for that...

File: ext/php8/span.h (hunk 7 above)
Comment (SammyK) [reply ext/php8/span.h:24]: Ah, makes sense. Thanks for clarifying.
